### PR TITLE
WIP: Add proxy support

### PR DESCRIPTION
**Pull request by micahlee**
### Desired Outcome

*Please describe the desired outcome for this PR.  Said another way, what was
the original request that resulted in these code changes?  Feel free to copy
this information from the connected issue.*

### Implemented Changes

*Describe how the desired outcome above has been achieved with this PR. In
particular, consider:*

- _What's changed? Why were these changes made?_
- _How should the reviewer approach this PR, especially if manual tests are required?_
- _Are there relevant screenshots you can add to the PR description?_

### Connected Issue/Story

Resolves #[relevant GitHub issue(s), e.g. 76]

CyberArk internal issue ID: [insert issue ID]

### Definition of Done
*At least 1 todo must be completed in the sections below for the PR to be
merged.*

#### Changelog

- [ ] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a
  CHANGELOG update

#### Test coverage

- [ ] This PR includes new unit and integration tests to go with the code
  changes, or
- [ ] The changes in this PR do not require tests

#### Documentation

- [ ] Docs (e.g. `README`s) were updated in this PR
- [ ] A follow-up issue to update official docs has been filed here: [insert issue ID]
- [ ] This PR does not require updating any documentation

#### Behavior

- [ ] This PR changes product behavior and has been reviewed by a PO, or
- [ ] These changes are part of a larger initiative that will be reviewed later, or
- [ ] No behavior was changed with this PR

#### Security

- [ ] Security architect has reviewed the changes in this PR,
- [ ] These changes are part of a larger initiative with a separate security review, or
- [ ] There are no security aspects to these changes

### CODE REVIEW

File: app/domain/authentication/authn_k8s/web_socket_client.rb [hunk 1]
```diff
@@ -124,17 +91,118 @@ def close
...
           send(nil, type: :close)
         end
+
         @closed = true
         @socket&.close
         @socket = nil
+
         emit(:__close)
+
         Thread.kill(@thread) if @thread
```
Comment (jvanderhoof) [inline app/domain/authentication/authn_k8s/web_socket_client.rb:101]: Is this to ensure we can safely spawn a new thread in `begin_event_loop` below? 

It feels a bit disjointed to have the lifecycle of a variable spread across three different classes, but without changing the interface, there's not much we can do.

**Review by jvanderhoof — APPROVED** [review]
Nice work on this @micahlee. I always like seeing what these protocols look like under the hood.

File: app/domain/authentication/authn_k8s/web_socket_client.rb (hunk 1 above)
Comment (micahlee) [reply app/domain/authentication/authn_k8s/web_socket_client.rb:101]: I don't think it so much for spawning a new thread, but just to make sure we clean up the thread already created.

I started down a path of refactoring more of the event loop, but I became concerned about verifying that it hadn't regressed in the timeframe we have for code freeze, so I backed out of that. There is definitely room to decouple some of these lifecycle and thread management routines, though. It's really difficult to try to add unit tests for this class the way it's written right now. The `_spec.rb` file for the websocket client is really more of an integration test right now. It gets the job done,  but is slower than it needs to be and doesn't really test all of the lifecycle edge cases.

Considering we've fixed 3+ bugs in this code over the past year, this wouldn't be a bad candidate for focusing on in a quality sprint at some point, to make this code easier to test and make changes to. I tried to do some of that with the OpenSSL setup and the proxy support, but the thread/event loop is still just kind of spaghetti code. CC: @adamouamani
